File: app/src/main/java/com/pep1lo/bakerspercentagecalculator/RecipeAdapter.java
package com.pep1lo.bakerspercentagecalculator;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.pep1lo.bakerspercentagecalculator.R;

import java.util.List;

public class RecipeAdapter extends RecyclerView.Adapter<RecipeAdapter.RecipeViewHolder> {

    private List<Recipe> recipes;
    private final OnItemClickListener clickListener;
    private final OnDeleteClickListener deleteClickListener;
    private final OnShareClickListener shareClickListener;

    public interface OnItemClickListener {
        void onItemClick(Recipe recipe);
    }

    public interface OnDeleteClickListener {
        void onDeleteClick(Recipe recipe);
    }

    public interface OnShareClickListener {
        void onShareClick(Recipe recipe, View anchor);
    }

    public RecipeAdapter(List<Recipe> recipes, OnItemClickListener clickListener, OnDeleteClickListener deleteClickListener, OnShareClickListener shareClickListener) {
        this.recipes = recipes;
        this.clickListener = clickListener;
        this.deleteClickListener = deleteClickListener;
        this.shareClickListener = shareClickListener;
    }

    @NonNull
    @Override
    public RecipeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_recipe, parent, false);
        return new RecipeViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecipeViewHolder holder, int position) {
        Recipe currentRecipe = recipes.get(position);
        holder.textViewRecipeName.setText(currentRecipe.getName());
        holder.itemView.setOnClickListener(v -> clickListener.onItemClick(currentRecipe));
        holder.buttonDeleteRecipe.setOnClickListener(v -> deleteClickListener.onDeleteClick(currentRecipe));
        holder.buttonShareRecipe.setOnClickListener(v -> shareClickListener.onShareClick(currentRecipe, v));
    }

    @Override
    public int getItemCount() {
        return recipes.size();
    }

    public void setRecipes(List<Recipe> recipes) {
        this.recipes = recipes;
        notifyDataSetChanged();
    }

    static class RecipeViewHolder extends RecyclerView.ViewHolder {
        final TextView textViewRecipeName;
        final ImageButton buttonDeleteRecipe;
        final ImageButton buttonShareRecipe;

        public RecipeViewHolder(@NonNull View itemView) {
            super(itemView);
            textViewRecipeName = itemView.findViewById(R.id.textViewRecipeName);
            buttonDeleteRecipe = itemView.findViewById(R.id.buttonDeleteRecipe);
            buttonShareRecipe = itemView.findViewById(R.id.buttonShareRecipe);
        }
    }
}

File: app/src/main/java/com/pep1lo/bakerspercentagecalculator/RecipeAdapter.kt
package com.pep1lo.bakerspercentagecalculator

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.pep1lo.bakerspercentagecalculator.databinding.ItemRecipeBinding

class RecipeAdapter(
    private var recipes: List<Recipe>,
    private val clickListener: (Recipe) -> Unit,
    private val deleteClickListener: (Recipe) -> Unit,
    private val shareClickListener: (Recipe, View) -> Unit
) : RecyclerView.Adapter<RecipeAdapter.RecipeViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecipeViewHolder {
        val binding = ItemRecipeBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return RecipeViewHolder(binding)
    }

    override fun onBindViewHolder(holder: RecipeViewHolder, position: Int) {
        val currentRecipe = recipes[position]
        holder.bind(currentRecipe)
    }

    override fun getItemCount(): Int = recipes.size

    fun setRecipes(recipes: List<Recipe>) {
        this.recipes = recipes
        notifyDataSetChanged()
    }

    inner class RecipeViewHolder(private val binding: ItemRecipeBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(recipe: Recipe) {
            binding.textViewRecipeName.text = recipe.name
            binding.root.setOnClickListener { clickListener(recipe) }
            binding.buttonDeleteRecipe.setOnClickListener { deleteClickListener(recipe) }
            binding.buttonShareRecipe.setOnClickListener { shareClickListener(recipe, it) }
        }
    }
}
